### ISSUE TITLE
rm -rf and override snapshots

### ISSUE THREAD

**Issue by j4jes** [openzfs/zfs#10485]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Centos
Distribution Version	| 7
Linux Kernel	|  3.10.0-957.el7.x86_64
Architecture	| 
ZFS Version	| 0.7.12-1.el7_6
SPL Version	| 0.7.12-1.el7_6

### Describe the problem you're observing
I'm wondering if it will ever be possible to rm -rf a file while telling it to not write changes to screenshots. I've read that snapshots are read only and absorb all changes to the volume since the creation of the snapshot, but in our case the movement of large files has become inevitable since we work with big datasets. We could enable deduplication but I'm not ready for that performance hit. Every time I do rm -rf to free up space I have to delete all our screen shots and since screenshots are a detrimental part of our work, this is living dangerously even with the presence of tape backups.

### Describe how to reproduce the problem
Have a ZFS volume with less than 5TB free space and someone moves a 2.5TB datafile to another working folder to run in a job, then deletes it when the job is done. No space left, have to remove all snapshots. Should of used symlinks right?

**Comment by gdevenyi** [comment]
I have the same issue and this is how I deal with it:
This sounds like a place where much more fine grained datasets come in. Users would have to be trained to know where "scratch" data goes, and where "important" data goes. My "scratch" ZFS filesystem has no snapshots, and users are educated that their data is not safe there. They are also educated that any "play" with processing that is not "final" goes in "scratch" until they're sure they want to save it for long-term.

**Comment by ahrens** [comment]
I think you're asking for a type of `rm` that would also remove the file from all snapshots that it's part of.  I don't know of anyone working on that.

**Comment by j4jes** [comment]
Yes I definitely could get more specific with my ZFS volumes (home, shared projects, external collaborators) and perhaps go for a scratch volume too. With the rm -rf question I think what I meant to say is to somehow have a delete function that works in tandem with rm. It wouldn't remove files from the snapshots since that is impossible, it would just throw data directly away without the snapshots claiming the changed bytes. Of course if such a thing were possible, then it could be incorporated for cp & mv commands too. I realize this would probably involve some complete overhauls of the way ZFS works wouldn't it? On some oblivious-to-the-pain, regular user level it seems like it would be a given for any snapshotting filesystem.

**Comment by kithrup** [comment]
Snapshots don't "claim the changed bytes"; the snapshots keep the original, unmodified value.

What you're asking for is not particularly possible with ZFS as it is currently implemented.  The only real workaround is to delete snapshots, or take fewer snapshots -- using more fine-grained datasets is one way to accomplish those.

**Comment by j4jes** [comment]
alright, I will reduce zfs snapshot retention period from two months to 1-2 weeks and rely on tape for anything older. This way deleted files should free up space sooner without needing to delete them all when I get in a crunch. I read somewhere that such a thing was possible with BTRFS, but perhaps the lack of strict guidelines is what lead to it's lack of integrity and stability.

**Comment by shodanshok** [comment]
I think this would be a clear reason to implement reflinks: simply `cp --reflink oldfile newfile` instead of plain coping it and eventual snapshot will only capture *changes* done to `newfile`, rather than the entire file. This it as a file-wide NOP write. 
@ahrens is reflink a feature which is planned in future releases?
